File: 2017/day18.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day18.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import parsefile

instr = parsefile(file_name, [[str], "\n"])


def is_num(s: str):
	if s[0] == "-":
		return s[1:].isdigit()
	return s.isdigit()


class VM:
	def __init__(self, instr):
		self.instr = [inst[:] for inst in instr]
		self.pc = 0
		self.registers = {}
		self.played_sounds = []

	def run(self):
		while 0 <= self.pc < len(self.instr):
			inst = self.instr[self.pc]

			if inst[0] == "snd":
				if is_num(inst[1]):
					value = int(inst[1])
				else:
					value = self.registers[inst[1]]

				self.played_sounds.append(value)
				self.pc += 1
			elif inst[0] == "set":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				self.registers[inst[1]] = value
				self.pc += 1
			elif inst[0] == "add":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				if inst[1] not in self.registers:
					self.registers[inst[1]] = 0

				self.registers[inst[1]] += value
				self.pc += 1
			elif inst[0] == "mul":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				if inst[1] not in self.registers:
					self.registers[inst[1]] = 0

				self.registers[inst[1]] *= value
				self.pc += 1
			elif inst[0] == "mod":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				if inst[1] not in self.registers:
					self.registers[inst[1]] = 0

				self.registers[inst[1]] %= value
				self.pc += 1
			elif inst[0] == "rcv":
				if is_num(inst[1]):
					value = int(inst[1])
				else:
					value = self.registers[inst[1]]

				if value != 0:
					return

				self.pc += 1
			elif inst[0] == "jgz":
				if is_num(inst[1]):
					value = int(inst[1])
				else:
					value = self.registers[inst[1]]

				if is_num(inst[2]):
					offset = int(inst[2])
				else:
					offset = self.registers[inst[2]]

				if value > 0:
					self.pc += offset
				else:
					self.pc += 1


def part1():
	vm = VM(instr)
	vm.run()

	return vm.played_sounds[-1]


class VM2:
	def __init__(self, instr, id, inputs=None):
		if inputs is None:
			inputs = []
		self.instr = [inst[:] for inst in instr]
		self.pc = 0
		self.registers = {"p": id}
		self.finished = False
		self.waiting_for_input = False

		if inputs is not None:
			self.inputs = inputs[:]
		else:
			self.inputs = []

		self.outputs = []
		self.send_count = 0
		self.id = id

	def run(self):
		while 0 <= self.pc < len(self.instr):
			inst = self.instr[self.pc]

			if inst[0] == "snd":
				if is_num(inst[1]):
					value = int(inst[1])
				else:
					value = self.registers[inst[1]]

				self.outputs.append(value)
				self.send_count += 1
				self.pc += 1
			elif inst[0] == "set":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				self.registers[inst[1]] = value
				self.pc += 1
			elif inst[0] == "add":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				if inst[1] not in self.registers:
					self.registers[inst[1]] = 0

				self.registers[inst[1]] += value
				self.pc += 1
			elif inst[0] == "mul":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				if inst[1] not in self.registers:
					self.registers[inst[1]] = 0

				self.registers[inst[1]] *= value
				self.pc += 1
			elif inst[0] == "mod":
				if is_num(inst[2]):
					value = int(inst[2])
				else:
					value = self.registers[inst[2]]

				if inst[1] not in self.registers:
					self.registers[inst[1]] = 0

				self.registers[inst[1]] %= value
				self.pc += 1
			elif inst[0] == "rcv":
				if len(self.inputs) != 0:
					self.registers[inst[1]] = self.inputs[0]
					self.inputs.pop(0)
				else:
					return

				self.pc += 1
			elif inst[0] == "jgz":
				if is_num(inst[1]):
					value = int(inst[1])
				else:
					value = self.registers[inst[1]]

				if is_num(inst[2]):
					offset = int(inst[2])
				else:
					offset = self.registers[inst[2]]

				if value > 0:
					self.pc += offset
				else:
					self.pc += 1

		self.finished = True

	def add_inputs(self, inputs):
		self.inputs.extend(inputs)
		if not self.finished:
			self.run()


def part2():
	vm0 = VM2(instr, 0)
	vm1 = VM2(instr, 1)

	while not vm0.finished or not vm1.finished:
		vm0.run()
		vm1.run()

		if len(vm0.outputs) == 0 and len(vm1.outputs) == 0:
			break

		vm0_out = vm0.outputs[:]
		vm0.outputs = []

		vm1_out = vm1.outputs[:]
		vm1.outputs = []

		vm0.add_inputs(vm1_out)
		vm1.add_inputs(vm0_out)

	return vm1.send_count


p1()
p2()

File: cpp/2017/day18.cpp
#include "../aocHelper.h"

struct VM_18_data
{
	enum class op_type
	{
		snd,
		set,
		add,
		mul,
		mod,
		rcv,
		jgz,
	};

	struct instr
	{
		op_type op;
		char reg1 = '\0';
		int value1 = 0;
		char reg2 = '\0';
		int value2 = 0;
	};
};

template<int type>
class VM_18
{
public:
	vector<VM_18_data::instr>& instructions;
	int id = 0;
	int pc = 0;
	unordered_map<char, long long> registers;
	bool finished = false;
	vector<int> played_sounds;
	int send_count = 0;
	deque<int> inputs;
	vector<int> outputs;

public:
	VM_18(vector<VM_18_data::instr>& instructions) : instructions(instructions)
	{
		static_assert(type == 0, "Use Other Constructor");

		for (int i = 'a'; i <= 'z'; i++)
		{
			registers[(char) i] = 0;
		}
	}

	VM_18(vector<VM_18_data::instr>& instructions, int id) : instructions(instructions)
	{
		static_assert(type == 1, "Use Other Constructor");
		this->id = id;

		for (int i = 'a'; i <= 'z'; i++)
		{
			registers[(char) i] = 0;
		}

		registers['p'] = id;
	}

	void run()
	{
		while (pc >= 0 && pc < instructions.size())
		{
			VM_18_data::instr& inst = instructions[pc];

			switch (inst.op)
			{
				case VM_18_data::op_type::snd:
				{
					if constexpr (type == 0)
					{
						played_sounds.push_back(registers[inst.reg1]);
					}
					else
					{
						int value;
						if (inst.reg1 == '\0')
						{
							value = inst.value1;
						}
						else
						{
							value = registers[inst.reg1];
						}

						outputs.push_back(value);
						send_count++;
					}

					pc++;
					break;
				}
				case VM_18_data::op_type::set:
				{
					int value;
					if (inst.reg2 == '\0')
					{
						value = inst.value2;
					}
					else
					{
						value = registers[inst.reg2];
					}

					registers[inst.reg1] = value;

					pc++;
					break;
				}
				case VM_18_data::op_type::add:
				{
					int value;
					if (inst.reg2 == '\0')
					{
						value = inst.value2;
					}
					else
					{
						value = registers[inst.reg2];
					}

					registers[inst.reg1] += value;

					pc++;
					break;
				}
				case VM_18_data::op_type::mul:
				{
					int value;
					if (inst.reg2 == '\0')
					{
						value = inst.value2;
					}
					else
					{
						value = registers[inst.reg2];
					}

					registers[inst.reg1] *= value;

					pc++;
					break;
				}
				case VM_18_data::op_type::mod:
				{
					int value;
					if (inst.reg2 == '\0')
					{
						value = inst.value2;
					}
					else
					{
						value = registers[inst.reg2];
					}

					registers[inst.reg1] %= value;

					pc++;
					break;
				}
				case VM_18_data::op_type::rcv:
				{
					if constexpr (type == 0)
					{
						int value;
						if (inst.reg1 == '\0')
						{
							value = inst.value1;
						}
						else
						{
							value = registers[inst.reg1];
						}

						if (value != 0)
						{
							return;
						}
					}
					else
					{
						if (inputs.size() > 0)
						{
							registers[inst.reg1] = inputs.front();
							inputs.pop_front();
						}
						else
						{
							return;
						}
					}

					pc++;
					break;
				}
				case VM_18_data::op_type::jgz:
				{
					int value;
					if (inst.reg1 == '\0')
					{
						value = inst.value1;
					}
					else
					{
						value = registers[inst.reg1];
					}

					int offset;
					if (inst.reg2 == '\0')
					{
						offset = inst.value2;
					}
					else
					{
						offset = registers[inst.reg2];
					}

					if (value > 0)
					{
						pc += offset;
					}
					else
					{
						pc++;
					}

					break;
				}
			}
		}

		finished = true;
	}

	void add_inputs(vector<int>& inputs)
	{
		static_assert(type == 1, "Function Only Allowed For Type 1");

		this->inputs.insert(this->inputs.end(), inputs.begin(), inputs.end());
		if (!finished)
		{
			run();
		}
	}
};

class Day18 : public BaseDay
{
public:
	Day18() : BaseDay("18") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		vector<VM_18_data::instr> instructions;

		// parse input
		while (*input != '\0')
		{
			VM_18_data::instr inst;

			if (*input == 's')
			{
				input++; // skip 's'

				if (*input == 'n') // snd X
				{
					inst.op = VM_18_data::op_type::snd;
					input += 3; // skip "nd "
					inst.reg1 = *input;
					input++;
				}
				else // set X Y
				{
					inst.op = VM_18_data::op_type::set;
					input += 3; // skip "et "
					inst.reg1 = *input;
					input += 2;
					if (*input >= 'a' && *input <= 'z')
					{
						inst.reg2 = *input;
						input++;
					}
					else
					{
						inst.value2 = numericParse<int>(input);
					}
				}
			}
			else if (*input == 'm')
			{
				input++; // skip 'm'
				if (*input == 'u') // mul X Y
				{
					inst.op = VM_18_data::op_type::mul;
					input += 3; // skip "ul "
					inst.reg1 = *input;
					input += 2;
					if (*input >= 'a' && *input <= 'z')
					{
						inst.reg2 = *input;
						input++;
					}
					else
					{
						inst.value2 = numericParse<int>(input);
					}
				}
				else // mod X Y
				{
					inst.op = VM_18_data::op_type::mod;
					input += 3; // skip "od "
					inst.reg1 = *input;
					input += 2;
					if (*input >= 'a' && *input <= 'z')
					{
						inst.reg2 = *input;
						input++;
					}
					else
					{
						inst.value2 = numericParse<int>(input);
					}
				}
			}
			else if (*input == 'a') // add X Y
			{
				inst.op = VM_18_data::op_type::add;
				input += 4; // skip "add "
				inst.reg1 = *input;
				input += 2;
				if (*input >= 'a' && *input <= 'z')
				{
					inst.reg2 = *input;
					input++;
				}
				else
				{
					inst.value2 = numericParse<int>(input);
				}
			}
			else if (*input == 'r') // rcv X
			{
				inst.op = VM_18_data::op_type::rcv;
				input += 4; // skip "rcv "
				inst.reg1 = *input;
				input++;
			}
			else if (*input == 'j') // jgz X Y
			{
				inst.op = VM_18_data::op_type::jgz;
				input += 4; // skip "jgz "
				if (*input >= 'a' && *input <= 'z')
				{
					inst.reg1 = *input;
					input++;
				}
				else
				{
					inst.value1 = numericParse<int>(input);
				}
				input++;
				if (*input >= 'a' && *input <= 'z')
				{
					inst.reg2 = *input;
					input++;
				}
				else
				{
					inst.value2 = numericParse<int>(input);
				}
			}

			instructions.push_back(inst);

			input++; // skip '\n'
		}

		// part 1
		VM_18<0> p1_vm{ instructions };
		p1_vm.run();
		part1 = p1_vm.played_sounds.back();

		// part 2
		VM_18<1> vm_0{ instructions, 0 };
		VM_18<1> vm_1{ instructions, 1 };

		while (!vm_0.finished || !vm_1.finished)
		{
			vm_0.run();
			vm_1.run();

			if (vm_0.outputs.size() == 0 && vm_1.outputs.size() == 0)
			{
				break;
			}

			vector<int> vm_0_out = vm_0.outputs;
			vm_0.outputs = {};

			vector<int> vm_1_out = vm_1.outputs;
			vm_1.outputs = {};

			vm_0.add_inputs(vm_1_out);
			vm_1.add_inputs(vm_0_out);
		}

		part2 = vm_1.send_count;

		return { part1, part2 };
	}
};
